core/git_remote.go

```diff
@@ -62,7 +67,132 @@ func (repo *RemoteGitRepository) PBDefinitions(ctx context.Context) ([]*pb.Defin
 	return nil, nil
 }
 
-func (repo *RemoteGitRepository) Remote(ctx context.Context) (*gitutil.Remote, error) {
+func (repo *RemoteGitRepository) Remote(ctx context.Context) (result *gitutil.Remote, rerr error) {
+	ctx, span := Tracer(ctx).Start(ctx, "git remote metadata", telemetry.Internal())
+	defer telemetry.End(span, func() error { return rerr })
+
+	slog := slog.SpanLogger(ctx, InstrumentationLibrary)
+
+	srv, err := CurrentDagqlServer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("query server: %w", err)
+	}
+	cacheKey := repo.remoteCacheKey()
+
+	if srv == nil || srv.Cache == nil {
+		slog.Info("git remote cache unavailable; running ls-remote", "cache_key", cacheKey)
+		return repo.runLsRemote(ctx)
+	}
+
+	cacheRes, err := srv.Cache.GetOrInitialize(ctx, enginecache.CacheKey[string]{
+		CallKey:        cacheKey,
+		ConcurrencyKey: cacheKey,
+	}, func(ctx context.Context) (dagql.AnyResult, error) {
+		remote, err := repo.runLsRemote(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		slog.Info("caching git remote metadata", "cache_key", cacheKey)
+		// Serialize to JSON so the cache can persist a plain string payload without
+		// requiring gitutil.Remote to satisfy dagql's typing interfaces. The decode
+		// step also hands each repo its own copy, so later tweaks (e.g. setting
+		// repo.Remote.Head) stay scoped to that caller.
+		payload, err := json.Marshal(remote)
+		if err != nil {
+			return nil, err
+		}
+
+		return dagql.NewResultForCurrentID(ctx, dagql.NewString(string(payload)))
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	slog.Info("loaded git remote metadata", "cache_hit", cacheRes.HitCache(), "cache_key", cacheKey)
+
+	val := cacheRes.Result()
+	strRes, ok := dagql.UnwrapAs[dagql.Result[dagql.String]](val)
+	if !ok {
+		return nil, fmt.Errorf("unexpected cache value type %T", val)
+	}
+
+	var remote gitutil.Remote
+	if err := json.Unmarshal([]byte(strRes.Self().String()), &remote); err != nil {
+		return nil, fmt.Errorf("decode cached remote: %w", err)
+	}
+	return &remote, nil
+}
+
+func (repo *RemoteGitRepository) Get(ctx context.Context, target *gitutil.Ref) (GitRefBackend, error) {
+	return &RemoteGitRef{
+		repo: repo,
+		Ref:  target,
+	}, nil
+}
+
+// remoteCacheKey mixes only the inputs that influence ls-remote so we avoid hashing
+// the entire repo object (no raw secret payloads or dagql wrapper state). We fold in
+// the remote URL, auth settings (usernames plus hashed secret/known-host metadata),
+// and service bindings (service digest + hostname + sorted aliases).
+func (repo *RemoteGitRepository) remoteCacheKey() string {
```

Comment:
I *think* this can be simplified to just include:
* The `SessionID` (as read from client metadata)
* the URL of the repo

At that point it will be scoped per-session (and have a lifetime limited to the session) and scoped per-URL. AFAIK the various auth-related things don't change the answer you get from `ls-remote` (just whether you get an error or not).

The current implementation actually has a corner case issue in that since it's not scoped to a session, concurrent clients will pick up this cache hit, which then creates a weird situation where we'll never re-run `ls-remote` again until no open client has a reference to it. So we definitely want to mix in `SessionID`, but at that point all the auth stuff should be unneeded too.

Reply:
After iterations and experiments with Tibor [see comment below](https://github.com/dagger/dagger/pull/11366#issuecomment-3503641972), our test suite (especially `TestGit`) was testing, in a single session, different patterns of how git operations fail after a successful run.

As the `git ls-remote` operation is being used to assess if a repo is public or private, I came to the conclusion that we need to include the auth sequence as part of the cache key.

Some real-use pipelines do exist and could be expressed as part of a CI / platform team pipelines. Example: Token rotation -> A job may fetch metadata with an expiring token, revoke it (or switch users), and immediately re-query to confirm the revocation took effect